[docs] Add "Migration" section to sidebar and revise v4-v5 content (#32740)
mui/material-ui: docs/data/material/migration/migration-v4/v5-style-changes.md

If you modify the shape of the theme after its creation, you need to migrate the structure manually.
:::

The following changes are supported by the adapter:

### Remove gutters

The "gutters" abstraction hasn't proven to be used frequently enough to be valuable.

```diff
-theme.mixins.gutters(),
+paddingLeft: theme.spacing(2),
+paddingRight: theme.spacing(2),
+[theme.breakpoints.up('sm')]: {
+  paddingLeft: theme.spacing(3),
+  paddingRight: theme.spacing(3),
+},
```

### ✅ Remove px suffix

`theme.spacing` now returns single values with px units by default.
This change improves the integration with styled-components & Emotion.

Before:

```js
theme.spacing(2) => 16
```

After:

```js
theme.spacing(2) => '16px'
```

### ✅ Rename theme.palette.type

The `theme.palette.type` key was renamed to `theme.palette.mode`, to better follow the "dark mode" terminology that is usually used for describing this feature.

```diff
 import { createTheme } from '@mui/material/styles';
-const theme = createTheme({palette: { type: 'dark' }}),
+const theme = createTheme({palette: { mode: 'dark' }}),
```

### Change default theme.palette.info colors

> The default `theme.palette.info` colors were changed to pass the AA standard contrast ratio in both light and dark modes.

Reviewer: For those that don't know what AA means?

```suggestion
The default `theme.palette.info` colors were changed to pass the AA accessibility standard contrast ratio in both light and dark modes.
```

x4 (the other colors)

+  main: green[800], // green[400] in "dark" mode

-  light: green[300],
+  light: green[500], // green[300] in "dark" mode

-  dark: green[700],
+  dark: green[900], // green[700] in "dark" mode
  }
```

### Change default theme.palette.warning colors

The default `theme.palette.warning` colors were changed to pass the AA standard contrast ratio in both light and dark modes.

```diff
  warning = {
-  main: orange[500],
+  main: '#ED6C02', // orange[400] in "dark" mode

-  light: orange[300],
+  light: orange[500], // orange[300] in "dark" mode

-  dark: orange[700],
+  dark: orange[900], // orange[700] in "dark" mode
  }
```

### Restore theme.palette.text.hint key (if needed)

The `theme.palette.text.hint` key was unused in Material UI components, and has been removed.
If you depend on it, you can add it back:

```diff
  import { createTheme } from '@mui/material/styles';

-const theme = createTheme(),
+const theme = createTheme({
+  palette: { text: { hint: 'rgba(0, 0, 0, 0.38)' } },
+});
```

### Restructure component definitions

The component definitions in the theme were restructured under the `components` key to make them easier to find.

#### 1. props

```diff

>   import { createTheme } from '@mui/material/styles';
> 
>   const theme = createTheme({

Reviewer: Wrong indent, it should only be one space for git diffs:

```suggestion
 import { createTheme } from '@mui/material/styles';

 const theme = createTheme({
```

-  props: {
-    MuiButton: {
-      disableRipple: true,
-    },
-  },
+  components: {
+    MuiButton: {
+      defaultProps: {
+        disableRipple: true,
+      },
+    },
+  },

>   });

Reviewer: ```suggestion
 });
```

```

#### 2. overrides

```diff
  import { createTheme } from '@mui/material/styles';

  const theme = createTheme({
-  overrides: {
-    MuiButton: {
-      root: { padding: 0 },
-    },
-  },
+  components: {
+    MuiButton: {
+      styleOverrides: {
+        root: { padding: 0 },
+      },
+    },
+  },

>   });

Reviewer: ```suggestion
 });
```

```ts
// in the file where you are creating the theme (invoking the function `createTheme()`)
import { Theme } from '@mui/material/styles';

declare module '@mui/styles' {
  interface DefaultTheme extends Theme {}
}
```

## @mui/material/colors

### ✅ Change color imports

Nested imports of more than one level are private. For example, you can no longer import `red` from `@mui/material/colors/red`.

```diff
-import red from '@mui/material/colors/red';
+import { red } from '@mui/material/colors';
```

## @mui/material/styles

### ✅ Rename fade to alpha

`fade` was renamed to `alpha` to better describe its functionality.

The previous name caused confusion when the input color already had an alpha value. The helper overrides the alpha value of the color.

```diff
-import { fade } from '@mui/material/styles';
+import { alpha } from '@mui/material/styles';

  const classes = makeStyles(theme => ({
-  backgroundColor: fade(theme.palette.primary.main, theme.palette.action.selectedOpacity),
+  backgroundColor: alpha(theme.palette.primary.main, theme.palette.action.selectedOpacity),
  }));
```

### ✅ Update createStyles import

> The `createStyles` function from `@mui/material/styles` was moved to the one exported from `@mui/styles`. It is necessary for removing the dependency to `@mui/styles` in the core package.

Reviewer: to be clearer?

```suggestion
The `createStyles` function from `@mui/material/styles` was moved to the one exported from `@mui/styles`. It is necessary for removing the dependency on `@mui/styles` in the Material UI npm package.
```

```diff
-import { jssPreset } from '@mui/material/styles';
+import { jssPreset } from '@mui/styles';
```

### ✅ Update `makeStyles` import

The `makeStyles` JSS utility is no longer exported from `@mui/material/styles`.
You can use `@mui/styles/makeStyles` instead.

Make sure to add a `ThemeProvider` at the root of your application, as the `defaultTheme` is no longer available.

If you are using this utility together with `@mui/material`, it's recommended that you use the `ThemeProvider` component from `@mui/material/styles` instead.

```diff
-import { makeStyles } from '@mui/material/styles';
+import { makeStyles } from '@mui/styles';
+import { createTheme, ThemeProvider } from '@mui/material/styles';

+const theme = createTheme();
  const useStyles = makeStyles((theme) => ({
    background: theme.palette.primary.main,
  }));
  function Component() {
    const classes = useStyles();
    return <div className={classes.root} />
  }

  // In the root of your app
  function App(props) {
-  return <Component />;
+  return <ThemeProvider theme={theme}><Component {...props} /></ThemeProvider>;
  }
```

> ### ✅ Update MuiThemeProvider import

Reviewer: I would suggest moving this diff up, to be right after `✅ Rename createMuiTheme` the two changes make sense together.
